crates/iceberg/src/metadata_scan.rs

```diff
@@ -0,0 +1,404 @@
+                DataType::Map(
+                    Arc::new(Field::new(
+                        "entries",
+                        DataType::Struct(
+                            vec![
+                                Field::new("keys", DataType::Utf8, false),
+                                Field::new("values", DataType::Utf8, true),
+                            ]
+                            .into(),
+                        ),
+                        false,
+                    )),
+                    false,
+                ),
+                false,
+            ),
+        ])
+    }
+
+    fn scan(scan: &MetadataScan) -> Result<RecordBatch> {
+        let mut committed_at =
+            PrimitiveBuilder::<TimestampMillisecondType>::new().with_timezone("+00:00");
+        let mut snapshot_id = PrimitiveBuilder::<Int64Type>::new();
+        let mut parent_id = PrimitiveBuilder::<Int64Type>::new();
+        let mut operation = StringBuilder::new();
+        let mut manifest_list = StringBuilder::new();
+        let mut summary = MapBuilder::new(None, StringBuilder::new(), StringBuilder::new());
+
+        for snapshot in scan.metadata_ref.snapshots() {
+            committed_at.append_value(snapshot.timestamp_ms());
+            snapshot_id.append_value(snapshot.snapshot_id());
+            parent_id.append_option(snapshot.parent_snapshot_id());
+            manifest_list.append_value(snapshot.manifest_list());
+            operation.append_value(snapshot.summary().operation.as_str());
+            for (key, value) in &snapshot.summary().additional_properties {
+                summary.keys().append_value(key);
+                summary.values().append_value(value);
+            }
+            summary.append(true)?;
+        }
+
+        Ok(RecordBatch::try_new(Arc::new(Self::schema()), vec![
+            Arc::new(committed_at.finish()),
+            Arc::new(snapshot_id.finish()),
+            Arc::new(parent_id.finish()),
+            Arc::new(operation.finish()),
+            Arc::new(manifest_list.finish()),
+            Arc::new(summary.finish()),
+        ])?)
+    }
+}
+
+/// Metadata log entries table.
+///
+/// Use to inspect the current and historical metadata files in the table.
+/// Contains every metadata file and the time it was added. For each metadata
+/// file, the table contains information about the latest snapshot at the time.
+pub struct MetadataLogEntriesTable;
+
+impl MetadataLogEntriesTable {
+    fn snapshot_id_as_of_time(
+        table_metadata: &TableMetadataRef,
+        timestamp_ms_inclusive: i64,
+    ) -> Option<&SnapshotRef> {
+        let mut snapshot_id = None;
+        // The table metadata snapshot log is chronological
+        for log_entry in table_metadata.history() {
+            if log_entry.timestamp_ms <= timestamp_ms_inclusive {
+                snapshot_id = Some(log_entry.snapshot_id);
+            }
+        }
+        snapshot_id.and_then(|id| table_metadata.snapshot_by_id(id))
+    }
```

Comment:
Linking [Java] and [Python] implementations to compute the latest snapshot given a timestamp.

[Java]: https://github.com/apache/iceberg/blob/1b5886d0e6a97043259b8cda18d146feb5a552a4/core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java#L330-L359
[Python]: https://github.com/apache/iceberg-python/blob/ab43c6ca5d844a90b30473fb1aa02b3ff34870ed/pyiceberg/table/__init__.py#L923-L933

Reply:
This assumes that `history()` is chronologically ordered because that gets asserted [`TableMetadata#try_normalize`](https://github.com/apache/iceberg-rust/blob/d51e818f9f48fbe7514c9e986a7583d48e0bc86c/crates/iceberg/src/spec/table_metadata.rs#L439).